engine/crates/engine-v2/schema/src/input_value/walker.rs

```diff
@@ -0,0 +1,123 @@
+use crate::{InputValue, RawInputValue, RawInputValueId, RawInputValues, SchemaInputValueId, SchemaWalker, StringId};
+
+/// Walker over a RawInputValue providing Serialize, Deserialize, Display and Into<InputValue<'_>>.
+#[derive(Clone, Copy)]
+pub struct RawInputValueWalker<'ctx, Ctx: RawInputValuesContext<'ctx>> {
+    pub(super) ctx: Ctx,
+    pub(super) value: &'ctx RawInputValue<Ctx::Str>,
+}
+
+impl<'ctx, Ctx> RawInputValueWalker<'ctx, Ctx>
+where
+    Ctx: RawInputValuesContext<'ctx>,
+{
+    pub(super) fn walk(&self, value: &'ctx RawInputValue<Ctx::Str>) -> Self {
+        Self { ctx: self.ctx, value }
+    }
+
+    /// Undefined values within InputObjects are silently removed according to the GraphQL spec.
+    /// But we can't remove an undefined argument
+    pub fn is_undefined(&self) -> bool {
+        match self.value {
+            RawInputValue::Undefined => true,
+            RawInputValue::Ref(id) => self.ctx.walk(*id).is_undefined(),
+            // I don't think this should ever happen, so should maybe be an unreachable!()
+            RawInputValue::SchemaRef(id) => self.ctx.schema_walk(*id).is_undefined(),
```

Comment:
I assume no, but just incase: is there ever a chance of `Ref` or `SchemaRef` becoming cyclical?

Reply:
Well the structure doesn't prevent it. With the current construction no. `Ref` is used as a variable placeholder and `SchemaRef` for default value. so fair point.